File: 2024/11/11.pl
#!/usr/bin/env swipl
% usage: swipl 11.pl [FILENAME]

:- module(aoc2024day11, []).
:- module(aoc2024day11).
:- initialization(main, main).
:- use_module(library(clpfd)).
:- use_module(library(dcg/basics)).
:- use_module(library(dcg/high_order)).

load(Result) --> sequence(integer, " ", Result), eol, eos.

split_digits(N, N1, N2) :-
    number_chars(N, C),
    length(C, Length),
    divmod(Length, 2, L, 0),
    length(C1, L),
    once(append(C1, C2, C)),
    number_chars(N1, C1),
    number_chars(N2, C2).

blink([], []).
blink([0 | S0], [1 | S]) :- !, blink(S0, S).
blink([N | S0], [N1, N2 | S]) :- split_digits(N, N1, N2), !, blink(S0, S).
blink([N | S0], [N1 | S]) :- N1 is N * 2024, blink(S0, S).

foldn(0, _, R, R) :- !.
foldn(N, Fn, R0, R) :-
    succ(N0, N),
    call(Fn, R0, R1),
    foldn(N0, Fn, R1, R).

part_one(Input, Result) :-
    foldn(25, blink, Input, Blinked),
    length(Blinked, Result).

main([File]) :-
    phrase_from_file(load(Input), File),
    part_one(Input, ResultOne),
    writeln(ResultOne).

main([]) :- main(['input.txt']).

File: 2024/11/11.jl
#!/usr/bin/env julia
# usage: julia 11.jl [FILENAME]

module AOC2024Day11
using Base.Iterators: flatmap

load(file::AbstractString) = parse.(Int, split(readline(file)))

fromdigits(ds::Vector{Int})::Int = foldr((b, a) -> 10a + b, ds)

cache1 = Dict{Int,Vector{Int}}(0 => [1])
function blink(stone::Int)::Vector{Int}
    get!(cache1, stone) do
        ds = digits(stone)
        l = length(ds)
        if iseven(l)
            [fromdigits(ds[1:l÷2]), fromdigits(ds[l÷2+1:end])]
        else
            [stone * 2024]
        end
    end
end

function ntimes(fn::Function, n::Int, init)
    for _ = 1:n
        init = flatmap(fn, init)
    end
    init
end

cache5 = Dict{Int,Vector{Int}}()

function blink5(stone::Int)::Vector{Int}
    get!(cache5, stone) do
        [ntimes(blink, 5, stone)...]
    end
end

cache25 = Dict{Int,Vector{Int}}()

function blink25(stone::Int)::Vector{Int}
    get!(cache25, stone) do
        [ntimes(blink5, 5, stone)...]
    end
end

blink75(stones::Vector{Int}) = ntimes(blink25, 3, stones)

partone(stones) = sum(length, blink25.(stones))
parttwo(stones) = sum(length, blink75(stones))

if abspath(PROGRAM_FILE) == @__FILE__
    input = load(get(ARGS, 1, joinpath(@__DIR__, "input.txt")))
    println(partone(input))
    println(parttwo(input))
end

end # module
